iceoryx_posh/source/popo/server_options.cpp

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "iceoryx_posh/popo/server_options.hpp"
+#include "iceoryx_posh/internal/log/posh_logging.hpp"
+
+namespace iox
+{
+namespace popo
+{
+cxx::Serialization ServerOptions::serialize() const noexcept
+{
+    return cxx::Serialization::create(requestQueueCapacity,
+                                      nodeName,
+                                      offerOnCreate,
+                                      static_cast<std::underlying_type_t<ConsumerTooSlowPolicy>>(clientTooSlowPolicy));
+}
+
+cxx::expected<ServerOptions, cxx::Serialization::Error>
+ServerOptions::deserialize(const cxx::Serialization& serialized) noexcept
+{
+    using ClientTooSlowPolicyUT = std::underlying_type_t<ConsumerTooSlowPolicy>;
+
+    ServerOptions serverOptions;
+    ClientTooSlowPolicyUT clientTooSlowPolicy;
+
+    auto deserializationSuccessful = serialized.extract(
+        serverOptions.requestQueueCapacity, serverOptions.nodeName, serverOptions.offerOnCreate, clientTooSlowPolicy);
+
+    if (!deserializationSuccessful
+        || clientTooSlowPolicy > static_cast<ClientTooSlowPolicyUT>(ConsumerTooSlowPolicy::DISCARD_OLDEST_DATA))
```

Comment:
Ok, this works since DISCARD_OLDEST_DATA is currently the max in the enum. The problem is when we extend the enum (carelessly by just appending a new value) but do forget to change the code here we do not correctly deserialize. 

We need to rely on tests catching the error for the new enum value then (but we need to add the tests with the changes and may forget to do so...). There is unfortunately no way to get the max of an enum or enum class I know of.

Some trick for enums to make it more manageable is 
```
enum class MyEnum {
          MIN=0, A=MIN, B, C, MAX = C };
```
This makes it more clear when extending with D that it is probably required to adapt MAX = D but it is also only weak protection and might not be worth it (someone may add D but forget changing MAX).

Also cf. https://here-be-braces.com/extending-cpp-enums/ , I do not know of a simple alternative, it appears there is none.

Reply:
Yes, it is not ideal. On the good side, if there is something added the deserialization of the new value would fail and we would immediately detect it in tests.

Reply:
I prefer the approach of adding an additional enum called `END` and then one can add additional enum values without touching any piece of code.

Reply:
This does not fully solve the problem since one can add new enum values after `END` and also leaks an implementation detail to the user since they would be able to configure `ConsumerTooSlowPolicy::END`. Usually it is not a good idea to have surrogates since they can also misused.

Since this is an enum class, one also has to edit the code when new values are added if there is a switch statement without a default value.